### ISSUE TITLE
Error: cannot use input image string (no discovered imageDigest)

### ISSUE THREAD

**Issue by staylorx** [anchore/anchore-cli#11]
I am having a problem with the anchore-cli running 'vuln'. It's entirely possible I just don't have the right signature, but I'm getting `Error: cannot use input image string (no discovered imageDigest)`.

I've added the image and it works as expected. 'image list' also. But using the server/namespace/tag format comes back with the error string. Using the sha256 has works properly but adds a step where I have to parse the json from the full output to get the hash, then resubmit. Bit of a pain, that.

Am I missing something simple?

$>anchore-cli image add docker.io/library/debian:latest

```
Image Digest: sha256:316ebb92ca66bb8ddc79249fb29872bece4be384cb61b5344fac4e84ca4ed2b2
Analysis Status: analyzed
Image Type: docker
Image ID: 2b98c9851a37e18356abe407922726dc35c370f7cee76f435f94ee92950b4493
Dockerfile Mode: Guessed
Distro: debian
Distro Version: 9
Size: 45135077
Architecture: amd64
Layer Count: 1

Full Tag: docker.io/library/debian:latest
```

Trying to pull the vulnerabilities using the full tag:

    $>anchore-cli image vuln docker.io/library/debian:latest os

yields

    Error: cannot use input image string (no discovered imageDigest)

Using the sha256 hash works as expected:

    $>anchore-cli --json image vuln sha256:316ebb92ca66bb8ddc79249fb29872bece4be384cb61b5344fac4e84ca4ed2b2 os

Thanks much.

**Comment by zhill** [comment]
Hi @staylorx, I tried the commands you listed on the latest code and it worked just fine, what version of anchore-engine are you running and which version of anchore-cli? That info will help me reproduce the issue. 

If you run: `anchore-cli image get docker.io/library/debian:latest` does it return the image record?

**Comment by staylorx** [comment]
Thanks for the reply.

Versions:
anchore-cli, version 0.1.10
anchore-engine: anchore-engine:v0.1.9 (Kubernetes via helm)

The image get command returned nothing.

     # anchore-cli image list

```
Full Tag                               Image ID                                                                Analysis
Status
docker.io/library/debian:latest        2b98c9851a37e18356abe407922726dc35c370f7cee76f435f94ee92950b4493        analyzed
docker.io/ubuntu:latest                c9d990395902a9e219684af847a63d793ccf72cf2aadeece1b576566c5662400        analyzed
```

Running `anchore-cli image get docker.io/library/debian:latest` brings back a new empty line.

**Comment by zhill** [comment]
Great, thanks. Can you run the same command with the '--debug' enabled?

`

**Comment by zhill** [comment]
Hi @staylorx are you running in GKE by any chance and accessing it via a LB or an Ingress?

**Comment by staylorx** [comment]
I'll run it with debug when I get back in front of my rig. In the meantime, yes @zhill I'm on GCP through an ingress with NodePort enabled. I had the the thought that maybe it was related because I'm coming into it via 80 and not 8228.

**Comment by zhill** [comment]
Good to know. I've run into this issue before on GCP. In GCP, the L7 load balancer it uses for Ingress definitions has a strange property where it won't pass the HTTP body on an HTTP GET request. The issue you're seeing is the due to the fact that our filter in the API for GET request on the image list is passed in the HTTP body (so we can do things like use full docker pull strings which aren't very URL friendly). In GCP that get stripped off and the query that hits the API service is missing the filter that the CLI uses to lookup the specific digest for that tag.

Can you try running the same command from inside the cluster using the pod IP directly? We have an anchore/engine-cli container you can use for that. The same command should work from within the cluster. Let me know if it doesn't and we can debug further.

**Comment by staylorx** [comment]
From the following I have to assume that what you say is true about something happening to the body of the request:

```
staylorx:~/workspace/src (master) $ anchore-cli --debug --json image get sha256:2441496fb9f0d938e5f8b27aba5cc367b24078225ceed82a9a5e67f0d6738c80                         
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 35.186.196.224
DEBUG:urllib3.connectionpool:http://35.186.196.224:80 "GET /v1 HTTP/1.1" 200 0
DEBUG:anchorecli.cli.image:discovery from input: imageDigest : sha256:2441496fb9f0d938e5f8b27aba5cc367b24078225ceed82a9a5e67f0d6738c80
DEBUG:anchorecli.clients.apiexternal:GET url=http://35.186.196.224/v1/images?history=false
DEBUG:anchorecli.clients.apiexternal:GET payload={"imageDigest": "sha256:2441496fb9f0d938e5f8b27aba5cc367b24078225ceed82a9a5e67f0d6738c80"}
DEBUG:anchorecli.clients.apiexternal:GET insecure=True
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 35.186.196.224
DEBUG:urllib3.connectionpool:http://35.186.196.224:80 "GET /v1/images?history=false HTTP/1.1" 400 1555
DEBUG:anchorecli.cli.utils:fetched httpcode from response: 400
"<!DOCTYPE html>\n<html lang=en>\n  <meta charset=utf-8>\n  <meta name=viewport content=\"initial-scale=1, minimum-scale=1, width=device-width\">\n  <title>Error 400 (Bad Request)!!1</title>\n  <style>\n    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}\n  </style>\n  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>\n  <p><b>400.</b> <ins>That\u2019s an error.</ins>\n  <p>Your client has issued a malformed or illegal request.  <ins>That\u2019s all we know.</ins>\n"
```

What's truly got me now is that it use to work at least with the full digest and now that's failing too.

I did try it with the anchore-cli command from within the cluster -- I had a custom image with the cli on it already -- and it seems to work fine against the internal network/CIDR.

**Comment by zhill** [comment]
Yes, that's the error you get from the L7  ALB that GCP sets up for the Ingress. An alternative is to use the LoadBalancer service type instead of NodePort + Ingress, but that will change how you configure TLS.

We're working on fixing things to not require GET body parsing for this reason, and I'll update this ticket when that gets merged and resolved.

**Comment by staylorx** [comment]
Thanks for the note. We moved past it with another solution. Please close this if you think it's fixed. I've got another Kubernetes upgrade install next month or so... and we'll check it out then.

**Comment by MMartyn** [comment]
@zhill I am hitting this using v0.2.3 of both engine and cli. I am able to add items to be analyzed, list the images, but once I try to view the vulnerabilities I get "Error: cannot use input image string (no discovered imageDigest)"

**Comment by nurmi** [comment]
hi @MMartyn -

Could you provide us with some output that illustrates the condition you're describing?  If an image has been added (say with 'anchore-cli image add docker.io/foobar:latest', then you should be able to use the same image identifier to access the vulnerability route 'anchore-cli image vuln docker.io/foobar:latest os') - if we can see the information and CLI/API calls you're making, perhaps we can see what is causing the inability of the engine to lookup the provided image.

**Comment by MMartyn** [comment]
Hi @nurmi, 

It seems like the same issue described above. Here are some debug outputs:

Version Info:
```
~ anchore-cli --debug system status          
...
Engine DB Version: 0.0.7
Engine Code Version: 0.2.3
```

```
~ anchore-cli --version
anchore-cli, version 0.2.3

```

Adding image:
```
~ anchore-cli --debug image add ubuntu:16.04                       
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): {ANCHORE_ENGINE_URL}
DEBUG:urllib3.connectionpool:https://{ANCHORE_ENGINE_URL}:443 "GET / HTTP/1.1" 200 0
DEBUG:anchorecli.clients.apiexternal:POST url=https://{ANCHORE_ENGINE_URL}/images?autosubscribe=True
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): {ANCHORE_ENGINE_URL}
DEBUG:urllib3.connectionpool:https://{ANCHORE_ENGINE_URL}:443 "POST /images?autosubscribe=True HTTP/1.1" 200 466
DEBUG:anchorecli.cli.utils:fetched httpcode from response: 200
Image Digest: sha256:0d06090fff94c0a3640729c7ef7e6b36ad5b41bec600cc2be92739c90d204243
Analysis Status: not_analyzed
Image Type: docker
Image ID: e13f3d529b1ad39b4f571b249de40754b7be4a63211c7c67be4addbd47839f2c
Dockerfile Mode: None
Distro: None
Distro Version: None
Size: None
Architecture: None
Layer Count: None

Full Tag: docker.io/ubuntu:16.04
```


Checking vulnerabilities:
```
~ anchore-cli --debug image vuln ubuntu:16.04
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): {ANCHORE_ENGINE_URL}
DEBUG:urllib3.connectionpool:https://{ANCHORE_ENGINE_URL}:443 "GET / HTTP/1.1" 200 0
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): swagger.json
Error: cannot use input image string (no discovered imageDigest)

```

**Comment by zhill** [comment]
@MMartyn, just to see, can you try `anchore-cli --debug image vuln docker.io/ubuntu:16.04` and see if you get the same result?

**Comment by MMartyn** [comment]
Same result:

 ```
 ~ anchore-cli --debug image vuln docker.io/ubuntu:16.04
  DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): {ANCHORE_ENGINE_URL}
  DEBUG:urllib3.connectionpool:https://{ANCHORE_ENGINE_URL}:443 "GET / HTTP/1.1" 200 0
  DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): swagger.json
  Error: cannot use input image string (no discovered imageDigest)
```

**Comment by zhill** [comment]
@MMartyn thanks. Can you try a few more?

`curl -u <user:pass> https://{ANCHORE_ENGINE_URL}/swagger.json`

That should return the swagger document from the apiext endpoint. That is used by the CLI client to detect the api version and toggle the GET behavior.

If that works, please also try:

`curl -u <user:pass> https://{ANCHORE_ENGINE_URL}/images/?fulltag=docker.io/ubuntu:16.04`

Example output from my system:
```
curl -u admin:foobar http://localhost:8228/images?fulltag=docker.io/alpine:latest
[
  {
    "analysis_status": "analyzed", 
    "analyzed_at": "2018-07-09T19:43:26Z", 
    "annotations": {}, 
    "created_at": "2018-07-09T19:37:46Z", 
    "imageDigest": "sha256:0873c923e00e0fd2ba78041bfb64a105e1ecb7678916d1f7776311e45bf5634b", 
    "image_content": {
      "metadata": {
        "arch": "amd64", 
        "distro": "alpine", 
        "distro_version": "3.8.0", 
        "dockerfile_mode": "Guessed", 
        "image_size": 2206542, 
        "layer_count": 1
      }
    }, 
    "image_detail": [
      {
        "created_at": "2018-07-18T22:36:42Z", 
        "digest": "sha256:0873c923e00e0fd2ba78041bfb64a105e1ecb7678916d1f7776311e45bf5634b", 
        "dockerfile": null, 
        "fulldigest": "docker.io/alpine@sha256:0873c923e00e0fd2ba78041bfb64a105e1ecb7678916d1f7776311e45bf5634b", 
        "fulltag": "docker.io/alpine:3.8", 
        "imageDigest": "sha256:0873c923e00e0fd2ba78041bfb64a105e1ecb7678916d1f7776311e45bf5634b", 
        "imageId": "11cd0b38bc3ceb958ffb2f9bd70be3fb317ce7d255c8a4c3f4af30e298aa1aab", 
        "last_updated": "2018-07-18T22:36:42Z", 
        "registry": "docker.io", 
        "repo": "alpine", 
        "tag": "3.8", 
        "tag_detected_at": "2018-07-18T22:36:42Z", 
        "userId": "admin"
      }
    ], 
    "image_status": "active", 
    "image_type": "docker", 
    "last_updated": "2018-07-09T19:43:26Z", 
    "userId": "admin"
  }
]
```

**Comment by zhill** [comment]
The fix for the url parsing, #17, is now in master branch but won't be pushed to pip until the next cli release.

**Comment by reyescameron17** [comment]
I seem to be having a similar problem with anchore-cli running both:
`anchore-cli image content docker.io/library/ubuntu:18.04` and `anchore-cli image vuln docker.io/library/ubuntu:18.04`
The image digest directly works fine:
```
docker-compose exec engine-api anchore-cli --json --debug image content docker.io/libary/ubuntu:18.04 os
DEBUG:anchorecli.clients.apiexternal:As Account = None
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): localhost:8228
DEBUG:urllib3.connectionpool:http://localhost:8228 "GET / HTTP/1.1" 200 5
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): localhost:8228
DEBUG:urllib3.connectionpool:http://localhost:8228 "GET /swagger.json HTTP/1.1" 200 180161
DEBUG:anchorecli.clients.apiexternal:As Account = None
DEBUG:anchorecli.clients.apiexternal:GET url=http://localhost:8228/images
DEBUG:anchorecli.clients.apiexternal:GET params={'fulltag': 'docker.io/libary/ubuntu:18.04', 'history': 'false'}
DEBUG:anchorecli.clients.apiexternal:Use get body because detected api version (0, 1, 14) < (0, 1, 6)? False
DEBUG:anchorecli.clients.apiexternal:GET insecure=True
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): localhost:8228
DEBUG:urllib3.connectionpool:http://localhost:8228 "GET /images?fulltag=docker.io%2Flibary%2Fubuntu%3A18.04&history=false HTTP/1.1" 404 106
DEBUG:anchorecli.cli.image:discovery from input: tag : docker.io/libary/ubuntu:18.04 : None
{
    "message": "cannot use input image string (no discovered imageDigest)"
}
```

I did an `image add` and `image wait` per the instructions.

```
docker-compose exec engine-api anchore-cli --version
anchore-cli, version 0.7.0
```

This is running in an EC2 instance with no LB or ingress. Simply an EIP mapped to it.

Other commands seem to work fine:

```
docker-compose exec engine-api anchore-cli image get docker.io/library/ubuntu:18.04
Image Digest: sha256:e5dd9dbb37df5b731a6688fa49f4003359f6f126958c9c928f937bec69836320
Parent Digest: sha256:bec5a2727be7fff3d308193cfde3491f8fba1a2ba392b7546b43a051853a341d
Analysis Status: analyzed
Image Type: docker
Analyzed At: 2020-04-02T16:09:07Z
Image ID: 4e5021d210f65ebe915670c7089120120bc0a303b90208592851708c1b8c04bd
Dockerfile Mode: Guessed
Distro: ubuntu
Distro Version: 18.04
Size: 69826560
Architecture: amd64
Layer Count: 4

Full Tag: docker.io/library/ubuntu:18.04
Tag Detected At: 2020-04-02T16:08:47Z
```
